File: packages/contracts/src/live.ts
import { z } from "zod";

import { SCHEMA_VERSION, schemaVersionField } from "./common.js";
import { SongRef } from "./song.js";

/**
 * Live, ephemeral signals broadcast during a running service (over Supabase
 * Realtime, on {@link liveChannel}). These are SIGNALS, not the source of truth:
 * the authoritative record (which song was shown, was_streamed) is written
 * separately as an idempotent {@link UsageEvent}. Stage publishes; Rec and any
 * other listener subscribe. `sequence` is a monotonic counter per service so a
 * late/duplicate signal can be ordered or dropped.
 */
const liveBase = {
  schema_version: schemaVersionField,
  service_id: z.string().uuid(),
  emitted_at: z.string(),
  sequence: z.number().int().min(0),
};

/** The presenter moved to a new cue/slide. */
export const CueAdvanced = z.object({
  ...liveBase,
  type: z.literal("cue.advanced"),
  item_id: z.string().nullable(),
  item_position: z.number().int().nullable(),
  label: z.string().nullable(),
  slide_index: z.number().int().nullable(),
});
export type CueAdvanced = z.infer<typeof CueAdvanced>;

/** A song became the active item — Rec can drop a chapter marker here. */
export const NowPlaying = z.object({
  ...liveBase,
  type: z.literal("now_playing"),
  song_ref: SongRef.nullable(),
  item_position: z.number().int().nullable(),
  title: z.string().nullable(),
});
export type NowPlaying = z.infer<typeof NowPlaying>;

/** The service went live (presentation started). */
export const ServiceLive = z.object({
  ...liveBase,
  type: z.literal("service.live"),
});
export type ServiceLive = z.infer<typeof ServiceLive>;

/** The service ended. */
export const ServiceEnded = z.object({
  ...liveBase,
  type: z.literal("service.ended"),
});
export type ServiceEnded = z.infer<typeof ServiceEnded>;

/** Any live signal, discriminated on `type`. */
export const LiveEvent = z.discriminatedUnion("type", [
  CueAdvanced,
  NowPlaying,
  ServiceLive,
  ServiceEnded,
]);
export type LiveEvent = z.infer<typeof LiveEvent>;

/** Fields every live-event builder needs: which service, when, and where in the
 * monotonic sequence. `sequence` is the per-service counter the caller advances
 * for each emit so listeners can order or drop late/duplicate signals. */
export interface LiveEventBase {
  serviceId: string;
  /** Monotonic per-service counter (0, 1, 2, …). */
  sequence: number;
  /** ISO 8601 UTC emit time; defaults to "now" so tests can pin it. */
  emittedAt?: string;
}

/** Build a validated {@link CueAdvanced} signal (presenter moved to a new cue). */
export function liveCueEvent(
  base: LiveEventBase,
  cue: {
    itemId?: string | null;
    itemPosition?: number | null;
    label?: string | null;
    slideIndex?: number | null;
  } = {},
): CueAdvanced {
  return CueAdvanced.parse({
    schema_version: SCHEMA_VERSION,
    type: "cue.advanced",
    service_id: base.serviceId,
    emitted_at: base.emittedAt ?? new Date().toISOString(),
    sequence: base.sequence,
    item_id: cue.itemId ?? null,
    item_position: cue.itemPosition ?? null,
    label: cue.label ?? null,
    slide_index: cue.slideIndex ?? null,
  });
}

/** Build a validated {@link NowPlaying} signal (a song became the active item). */
export function nowPlayingEvent(
  base: LiveEventBase,
  song: {
    songRef?: SongRef | null;
    itemPosition?: number | null;
    title?: string | null;
  } = {},
): NowPlaying {
  return NowPlaying.parse({
    schema_version: SCHEMA_VERSION,
    type: "now_playing",
    service_id: base.serviceId,
    emitted_at: base.emittedAt ?? new Date().toISOString(),
    sequence: base.sequence,
    song_ref: song.songRef ?? null,
    item_position: song.itemPosition ?? null,
    title: song.title ?? null,
  });
}

File: crates/sunday-contracts/src/live.rs
use serde::{Deserialize, Serialize};

use crate::common::{default_schema_version, SCHEMA_VERSION};
use crate::song::SongRef;

/// Live, ephemeral signals broadcast during a running service (Supabase Realtime,
/// on [`crate::live_channel`]). SIGNALS, not source of truth — the authoritative
/// record is a separate idempotent [`crate::UsageEvent`]. `sequence` is a
/// monotonic per-service counter for ordering/de-duplication. Internally tagged
/// on `type`.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum LiveEvent {
    /// The presenter moved to a new cue/slide.
    #[serde(rename = "cue.advanced")]
    CueAdvanced {
        #[serde(default = "default_schema_version")]
        schema_version: u32,
        service_id: String,
        emitted_at: String,
        sequence: u64,
        item_id: Option<String>,
        item_position: Option<i64>,
        label: Option<String>,
        slide_index: Option<i64>,
    },
    /// A song became the active item — Rec can drop a chapter marker here.
    #[serde(rename = "now_playing")]
    NowPlaying {
        #[serde(default = "default_schema_version")]
        schema_version: u32,
        service_id: String,
        emitted_at: String,
        sequence: u64,
        song_ref: Option<SongRef>,
        item_position: Option<i64>,
        title: Option<String>,
    },
    /// The service went live (presentation started).
    #[serde(rename = "service.live")]
    ServiceLive {
        #[serde(default = "default_schema_version")]
        schema_version: u32,
        service_id: String,
        emitted_at: String,
        sequence: u64,
    },
    /// The service ended.
    #[serde(rename = "service.ended")]
    ServiceEnded {
        #[serde(default = "default_schema_version")]
        schema_version: u32,
        service_id: String,
        emitted_at: String,
        sequence: u64,
    },
}

impl LiveEvent {
    /// Build a [`LiveEvent::CueAdvanced`] signal. `sequence` is the monotonic
    /// per-service counter; `emitted_at` is supplied by the caller (no clock
    /// dependency). Mirrors the TypeScript `liveCueEvent`.
    #[allow(clippy::too_many_arguments)]
    pub fn cue_advanced(
        service_id: String,
        sequence: u64,
        emitted_at: String,
        item_id: Option<String>,
        item_position: Option<i64>,
        label: Option<String>,
        slide_index: Option<i64>,
    ) -> Self {
        LiveEvent::CueAdvanced {
            schema_version: SCHEMA_VERSION,
            service_id,
            emitted_at,
            sequence,
            item_id,
            item_position,
            label,
            slide_index,
        }
    }

    /// Build a [`LiveEvent::NowPlaying`] signal. Mirrors the TypeScript
    /// `nowPlayingEvent`.
    pub fn now_playing(
        service_id: String,
        sequence: u64,
        emitted_at: String,
        song_ref: Option<SongRef>,
        item_position: Option<i64>,
        title: Option<String>,
    ) -> Self {
        LiveEvent::NowPlaying {
            schema_version: SCHEMA_VERSION,
            service_id,
            emitted_at,
            sequence,
            song_ref,
            item_position,
            title,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn cue_advanced_builder_sets_sequence_and_type() {
        let e = LiveEvent::cue_advanced(
            "svc".to_string(),
            5,
            "2026-05-31T09:01:00Z".to_string(),
            Some("item-2".to_string()),
            Some(2),
            Some("Verse 1".to_string()),
            Some(0),
        );
        match e {
            LiveEvent::CueAdvanced {
                schema_version,
                sequence,
                label,
                ..
            } => {
                assert_eq!(schema_version, SCHEMA_VERSION);
                assert_eq!(sequence, 5);
                assert_eq!(label.as_deref(), Some("Verse 1"));
            }
            _ => panic!("expected CueAdvanced"),
        }
    }

    #[test]
    fn now_playing_builder_defaults_to_none() {
        let e = LiveEvent::now_playing(
            "svc".to_string(),
            1,
            "2026-05-31T09:00:00Z".to_string(),
            None,
            None,
            None,
        );
        match e {
            LiveEvent::NowPlaying {
                sequence,
                song_ref,
                item_position,
                ..
            } => {
                assert_eq!(sequence, 1);
                assert!(song_ref.is_none());
                assert!(item_position.is_none());
            }
            _ => panic!("expected NowPlaying"),
        }
    }
}
